### PR TITLE
BREAKING CHANGE: Brightcove video launcher

### PR DESCRIPTION
**Pull request by dickie81**
This is a component to wrap the brightcove-video component. It presents the user with the videos poster with a play button overlaid. When this is clicked it launches the video with autoplay enabled.

The purpose of this is to stop initialising the full video to improve performance in scrolling lists etc.

### CODE REVIEW

**Comment by dickie81** [comment]
launcher

File: packages/brightcove-video-launcher/brightcove-video-launcher.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  }
+}
+
+BrightcoveVideoLauncher.propTypes = Object.assign(
+  {
+    poster: Image.propTypes.source,
+    playIcon: Image.propTypes.source,
+    playIconWidth: PropTypes.number,
+    playIconHeight: PropTypes.number
+  },
```
[the remark this answers is not in the document]
Comment (jackharvey1) [reply packages/brightcove-video-launcher/brightcove-video-launcher.js]: We've changed the icons as part of the design changes

[the remark this answers is not in the document]
Comment (dickie81) [reply packages/brightcove-video-launcher/brightcove-video-launcher.js]: I was trying to make it flexible.

File: packages/brightcove-video-launcher/brightcove-video-launcher.test.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+});
+
+it("renders poster with custom play icon if specified", () => {
+  const tree = renderer
+    .create(
+      <BrightcoveVideoLauncher
+        policyKey={policyKey}
+        videoId={videoId}
+        accountId={accountId}
+        playIcon={customPlayIcon}
```
[the remark this answers is not in the document]
Comment (dickie81) [reply packages/brightcove-video-launcher/brightcove-video-launcher.test.js]: flexibility

File: packages/brightcove-video/tap-to-launch.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import React, { Component } from "react";
+import { TouchableWithoutFeedback } from "react-native";
+import PropTypes from "prop-types";
+
+class TapToLaunch extends Component {
```
Comment (aronshamash) [inline packages/brightcove-video/tap-to-launch.js]: Not sure this is so great having a `TapToLaunch` component inside the brightcove package ?

Comment (dickie81) [reply packages/brightcove-video/tap-to-launch.js]: Can split it out easily enough. Just don't want to make assumptions. @trevorah what is your preference?

File: packages/brightcove-video/play-icon.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import React from "react";
```
Comment (jackharvey1) [inline packages/brightcove-video/play-icon.js]: I think we should name things `brightcove-...` if they're not generalised

File: packages/brightcove-video/tap-to-launch.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import React, { Component } from "react";
```
Comment (jackharvey1) [inline packages/brightcove-video/tap-to-launch.js]: What do we think about generalising this and pulling it out of `brightcove-video/`?

My idea had been a component that accepted four parameters: an unlaunched component and its props, and the launched component and its props.

File: packages/brightcove-video/tap-to-launch.js (hunk 3 above)
Comment (trevorah) [reply packages/brightcove-video/tap-to-launch.js]: I prefer to only creating a new package for a component if more than one thing is going to use it. This gives us more slack to refactor it without making a breaking change.

File: packages/brightcove-video/tap-to-launch.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    this.launch = this.launch.bind(this);
+  }
+
+  launch() {
+    this.setState({ isLaunched: true });
+  }
+
+  render() {
+    return (
+      <TouchableWithoutFeedback onPress={this.launch}>
```
Comment (trevorah) [inline packages/brightcove-video/tap-to-launch.js]: simplify with
```js
onPress={() => this.setState({ isLaunched: true })}
```

File: packages/brightcove-video/splash.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  playIconHeight: 70
+};
+Splash.propTypes = {
+  poster: Image.propTypes.source,
+  playIcon: Image.propTypes.source,
+  playIconWidth: PropTypes.number,
+  playIconHeight: PropTypes.number,
+  width: PropTypes.number.isRequired,
+  height: PropTypes.number.isRequired
+};
```
Comment (trevorah) [inline packages/brightcove-video/splash.js]: If you're going to allow for custom play icons, why limit yourself to images sources?

Have `playIcon` be a generic component, with your own PlayIcon component (with 70 as the width/height) as the default. Your styling is good enough to center whatever anyone passes in.

If someone wants to pass in
```
<Text>▶️</Text>
```
then they can go nuts.

File: packages/brightcove-video/brightcove-video.stories.js [hunk 8]
```diff
@@ -1,156 +1,33 @@
...
-  .add("With autoplay & event listner", () =>
+  .add("with custom play icon", () =>
     <BrightcoveVideo
       policyKey={policyKey}
       videoId={videoId}
       accountId={accountId}
-      onChange={firstArgJSONAction("change")}
-      autoplay
+      poster={{ uri: posterImageURI }}
+      playIcon={customPlayIcon}
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.stories.js]: still eww.

File: packages/brightcove-video/brightcove-video.stories.js [hunk 9]
```diff
@@ -1,156 +1,33 @@
...
-    />
-  )
-  .add("With external controls", () =>
-    <VideoWithExternalControls
-      policyKey={policyKey}
-      videoId={videoId}
-      accountId={accountId}
-    />
-  )
-  .add("With autoplay", () =>
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.stories.js]: Why the split in stories?

File: packages/brightcove-video/brightcove-video.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import Splash from "./splash";
+import TapToLaunch from "./tap-to-launch";
+
+const BrightcoveVideo = props =>
+  <TapToLaunch>
+    {isLaunched =>
+      <View style={{ width: props.width, height: props.height }}>
+        {isLaunched ? <Player {...props} /> : <Splash {...props} />}
+      </View>}
+  </TapToLaunch>;
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.js]: So how would autoplay work if the splash screen is always shown?

File: packages/brightcove-video/splash.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { View, Image } from "react-native";
+import TimesImage from "@times-components/image";
```
Comment (trevorah) [inline packages/brightcove-video/splash.js]: Why are you using this rather than the regular react native image component?

**Comment by jackharvey1** [comment]
I've just realised this is a breaking change as commonJS has to require `@times-components/brightcove-player/brightcove-player.web`, but we only have `brightcove-player`

File: packages/brightcove-video/tap-to-launch.js (hunk 6 above)
Comment (dickie81) [reply packages/brightcove-video/tap-to-launch.js]: facebook docs recommend against this:
https://facebook.github.io/react/docs/handling-events.html

Last paragraph

File: packages/brightcove-video/brightcove-video.stories.js (hunk 8 above)
Comment (dickie81) [reply packages/brightcove-video/brightcove-video.stories.js]: could you elaborate?

File: packages/brightcove-video/brightcove-video.js (hunk 10 above)
Comment (dickie81) [reply packages/brightcove-video/brightcove-video.js]: fair point

File: packages/brightcove-video/play-icon.js (hunk 4 above)
Comment (dickie81) [reply packages/brightcove-video/play-icon.js]: @trevorah?

**Pull request merged**